### PR TITLE
Pull request by Avi Dixit

### PR DESCRIPTION
**Pull request by addt**

### CODE REVIEW

File: D04_ex00.py [hunk 1]
```diff
@@ -16,14 +16,35 @@
 # Body
 
 
+def random():
+ import random
+ random_no = random.randrange(1,25,1)
```
Comment (danielsgriffin) [inline D04_ex00.py]: `random.randrange(1,25,1)` => + `random.randrange(1, 25, 1)`

Note: the step of 1 is not required.

File: D04_ex00.py [hunk 2]
```diff
@@ -16,14 +16,35 @@
 # Body
 
 
+def random():
+ import random
```
Comment (danielsgriffin) [inline D04_ex00.py]: generally advised to do outside function at top of script

File: HW04_ch07_ex02.py [hunk 3]
```diff
@@ -24,12 +24,25 @@
 
 
 # Body
-
+def eval_loop():
```
Comment (danielsgriffin) [inline HW04_ch07_ex02.py]: Good. Only change: "**return** the value of the last expression it evaluated"

File: HW04_ch08_ex04.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 ###############################################################################
 # Body
 
-
+#Correct
```
Comment (danielsgriffin) [inline HW04_ch08_ex04.py]: Incorrect.

File: HW04_ch08_ex05.py [hunk 5]
```diff
@@ -2,3 +2,15 @@
 # See Chapter 8: Strings Exercise 5 for the task.
 # Please do provide function calls that test/demonstrate your
```
Comment (danielsgriffin) [inline HW04_ch08_ex05.py]:
```
Book Examples:
  X  tried: 'A' got: 'x' expected: 'D'
  X  tried: 'Z' got: 'u' expected: 'A'
 OK  tried: 'cheer' got: 'jolly' expected: 'jolly'
 OK  tried: 'melon' got: 'cubed' expected: 'cubed'
  X  tried: 'IBM' got: 'buf' expected: 'HAL'
Daniel's Tests:
 OK  tried: 'abc' got: 'bcd' expected: 'bcd'
 OK  tried: 'xyz' got: 'yza' expected: 'yza'
  X  tried: 'ABC' got: 'vwx' expected: 'BCD'
  X  tried: 'XYZ' got: 'stu' expected: 'YZA'
  X  tried: 'Az' got: 'va' expected: 'Ba'
  X  tried: 'Aa' got: 'ua' expected: 'Aa'
 OK  tried: 'abc' got: 'zab' expected: 'zab'
 OK  tried: 'xyz' got: 'wxy' expected: 'wxy'
  X  tried: 'ABC' got: 'tuv' expected: 'ZAB'
  X  tried: 'XYZ' got: 'qrs' expected: 'WXY'
  X  tried: 'Az' got: 'ty' expected: 'Zy'
  X  tried: 'Aa' got: 'ua' expected: 'Aa'
```

**Comment by danielsgriffin** [comment]
Good job. Pay attention to your tests.

File: HW04_ch08_ex05.py [hunk 6]
```diff
@@ -2,3 +2,15 @@
...
 # function.
+
+def rotate(s,n):
```
Comment (danielsgriffin) [inline HW04_ch08_ex05.py]: rotate_word

File: HW04_string2.py [hunk 7]
```diff
@@ -41,9 +50,9 @@ def not_bad(s):
 # e.g. 'abcde', the front half is 'abc', the back half 'de'.
 # Given 2 strings, a and b, return a string of the form
 #  a-front + b-front + a-back + b-back
+import math
```
Comment (ghvik) [inline HW04_string2.py]: Perhaps use:`from math import ceil`, so you don't have to repeat math.ceil below

**Pull request closed without merging**